lib/spack/spack/operating_systems/cray_backend.py

```diff
@@ -95,8 +98,7 @@ def _detect_crayos_version(cls):
             v = read_clerelease_file()
             return spack.version.Version(v)[0]
         else:
-            raise spack.error.UnsupportedPlatformError(
-                'Unable to detect Cray OS version')
+            return None
```

Comment:
this may lead to hard to debug errors if that ever crops up somewhere, what is the reason this can't raise an error?

Reply:
Because not all Cray systems run CNL on the backend. We have a system at Livermore running in what Cray calls "cluster" mode that runs RHEL7 under the Cray PE.

So if we don't detect any Cray OS version on the system, we return `None`. We can't ever be sure we will get a Cray OS version. I thought this was cleaner than a `try/except` block in the client code.

Reply:
given that exceptions are first-class citizens in Python, personally I would rather ensure that `__str__` returns a string and raises an exception if it is unable to do so

Reply:
`str` already ensures that it returns a string, this is in `_detect_crayos_version`.

Reply:
This also needs a comment describing what you put in the comment above.